pkg/networkservice/common/heal/client.go

```diff
@@ -19,41 +19,208 @@ package heal
 
 import (
 	"context"
+	"sync"
+	"time"
 
+	"github.com/edwarnicke/serialize"
 	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/pkg/errors"
 	"google.golang.org/grpc"
 
 	"github.com/networkservicemesh/api/pkg/api/networkservice"
-
 	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/log"
 )
 
-// RegisterClientFunc - required to inform heal server about new client connection and assign it to the connection ID
-type RegisterClientFunc func(context.Context, *networkservice.Connection, networkservice.MonitorConnectionClient)
-
 type healClient struct {
-	ctx context.Context
-	cc  networkservice.MonitorConnectionClient
+	ctx              context.Context
+	cc               networkservice.MonitorConnectionClient
+	initOnce         sync.Once
+	initErr          error
+	connCache        map[string]*networkservice.Connection
+	conCacheExecutor serialize.Executor
+	expectedConns    verificationRequestMap
 }
 
 // NewClient - creates a new networkservice.NetworkServiceClient chain element that inform healServer about new client connection
 func NewClient(ctx context.Context, cc networkservice.MonitorConnectionClient) networkservice.NetworkServiceClient {
 	return &healClient{
-		ctx: ctx,
-		cc:  cc,
+		ctx:       ctx,
+		cc:        cc,
+		connCache: make(map[string]*networkservice.Connection),
 	}
 }
 
 func (u *healClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
+	u.initOnce.Do(func() {
+		errCh := make(chan error, 1)
+		pushFunc := healRequestFunc(ctx)
+		go u.listenToConnectionChanges(pushFunc, request.GetConnection().GetCurrentPathSegment(), errCh)
+		u.initErr = <-errCh
+	})
+	// if initialization failed, then we want for all subsequent calls to Request() on this object to also fail
+	if u.initErr != nil {
+		return nil, u.initErr
+	}
+
+	verificationRequest := u.addToExpectedConnections(request.GetConnection())
+	if verificationRequest != nil {
+		defer close(verificationRequest.stopCh)
+	}
+
 	conn, err := next.Client(ctx).Request(ctx, request, opts...)
+	if err != nil {
+		return nil, err
+	}
 
-	registerClient := registerClientFunc(ctx)
-	if err == nil && registerClient != nil {
-		registerClient(u.ctx, conn, u.cc)
+	if verificationRequest != nil {
+		err = u.confirmConnectionSuccess(conn, verificationRequest)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		u.addConnectionToMonitor(conn)
 	}
-	return conn, err
+
+	return conn, nil
 }
 
 func (u *healClient) Close(ctx context.Context, conn *networkservice.Connection, opts ...grpc.CallOption) (*empty.Empty, error) {
+	if cachedConn, ok := u.removeConnectionFromMonitor(conn); ok {
+		conn = cachedConn
+	} else {
+		log.FromContext(ctx).WithField("healClient", "Close").Warnf("can't find connection in cache, id=%v", conn.GetId())
```

Comment:
No entry in cache means no connection exists, here we can simply return `new(empty.Empty), nil`.